source/common/network/udp_listener_impl.cc

```diff
@@ -94,41 +68,61 @@ void UdpListenerImpl::onSocketEvent(short flags) {
   ENVOY_UDP_LOG(trace, "handleReadCallback");
-  sockaddr_storage addr;
-  socklen_t addr_len = 0;
-
+  // Max UDP payload.
+  constexpr uint64_t const read_length = 1500;
   do {
-    ReceiveResult recv_result = doRecvFrom(addr, addr_len);
-    if ((recv_result.result_.rc_ < 0)) {
-      if (recv_result.result_.errno_ != EAGAIN) {
-        ENVOY_UDP_LOG(error, "recvfrom result {}", recv_result.result_.errno_);
+    Address::InstanceConstSharedPtr peer_address;
+    Address::InstanceConstSharedPtr local_address;
+    Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+    Buffer::RawSlice slice;
+    const uint64_t num_slices = buffer->reserve(read_length, &slice, 1);
+    ASSERT(num_slices == 1);
+
+    uint32_t old_packets_dropped = packets_dropped_;
+    MonotonicTime receive_time = time_source_.monotonicTime();
+    Api::IoCallUint64Result result = socket_.ioHandle().recvmsg(
+        &slice, num_slices, socket_.localAddress()->ip()->port(),
+        /*v6only=*/socket_.localAddress()->ip()->version() == Address::IpVersion::v6,
+        &packets_dropped_, local_address, peer_address);
+
+    if (!result.ok()) {
+      // No more to read or encountered a system error.
+      if (result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+        ENVOY_UDP_LOG(error, "recvfrom result {}: {}",
+                      static_cast<int>(result.err_->getErrorCode()),
+                      result.err_->getErrorDetails());
         cb_.onReceiveError(UdpListenerCallbacks::ErrorCode::SyscallError,
-                           recv_result.result_.errno_);
+                           result.err_->getErrorCode());
       }
+      // Stop reading.
       return;
     }
 
-    if (recv_result.result_.rc_ == 0) {
+    if (result.rc_ == 0) {
       // TODO(conqerAtapple): Is zero length packet interesting?
-      return;
+      ENVOY_UDP_LOG(warn, "received 0-length packet");
     }
 
-    Address::InstanceConstSharedPtr local_address = socket_.localAddress();
+    if (local_address == nullptr) {
+      ENVOY_UDP_LOG(error, "fail to get local address from IP header");
+    }
 
-    RELEASE_ASSERT(
-        addr_len > 0,
-        fmt::format(
-            "Unable to get remote address for fd: {}, local address: {}. address length is 0 ",
-            socket_.ioHandle().fd(), local_address->asString()));
+    if (packets_dropped_ != old_packets_dropped) {
+      // The kernel tracks SO_RXQ_OVFL as a uint32 which can overflow to a smaller
+      // value. So as long as this count differs from previously recorded value,
+      // more packets are dropped by kernel.
+      uint32_t delta = (packets_dropped_ > old_packets_dropped)
+                           ? (packets_dropped_ - old_packets_dropped)
+                           : (packets_dropped_ +
+                              (std::numeric_limits<uint32_t>::max() - old_packets_dropped) + 1);
+      ENVOY_UDP_LOG(warn, "Kernel dropped {} more packets.", delta);
```

Comment:
I'm not sure I follow. Are you saying that the kernel dropped more packets then we are accounting for? It seems like this can happen and it shouldn't be a warn but actually debug with stat eventually? Can we add a TODO/comment?